crates/eframe/src/web/text_agent.rs

```diff
@@ -1,237 +1,177 @@
-        })?;
+        };
 
-        runner_ref.add_event_listener(
-            &input,
-            "compositionupdate",
+        let on_composition_update = {
             move |event: web_sys::CompositionEvent, runner: &mut AppRunner| {
-                if let Some(text) = event.data() {
-                    let egui_event = egui::Event::Ime(egui::ImeEvent::Preedit(text));
-                    runner.input.raw.events.push(egui_event);
-                    runner.needs_repaint.repaint_asap();
-                }
-            },
-        )?;
-
-        runner_ref.add_event_listener(&input, "compositionend", {
-            let input_clone = input.clone();
+                let Some(text) = event.data() else { return };
+                let event = egui::Event::Ime(egui::ImeEvent::Preedit(text));
+                runner.input.raw.events.push(event);
+                runner.needs_repaint.repaint_asap();
+            }
+        };
 
+        let on_composition_end = {
+            let input = input.clone();
             move |event: web_sys::CompositionEvent, runner: &mut AppRunner| {
-                is_composing.set(false);
-                input_clone.set_value("");
-
-                if let Some(text) = event.data() {
-                    let egui_event = egui::Event::Ime(egui::ImeEvent::Commit(text));
-                    runner.input.raw.events.push(egui_event);
-                    runner.needs_repaint.repaint_asap();
-                }
+                let Some(text) = event.data() else { return };
+                input.set_value("");
+                let event = egui::Event::Ime(egui::ImeEvent::Commit(text));
+                runner.input.raw.events.push(event);
+                runner.needs_repaint.repaint_asap();
             }
-        })?;
-    }
+        };
 
-    // When input lost focus, focus on it again.
-    // It is useful when user click somewhere outside canvas.
-    let input_refocus = input.clone();
-    runner_ref.add_event_listener(&input, "focusout", move |_event: web_sys::MouseEvent, _| {
-        // Delay 10 ms, and focus again.
-        let input_refocus = input_refocus.clone();
-        call_after_delay(std::time::Duration::from_millis(10), move || {
-            input_refocus.focus().ok();
-        });
-    })?;
+        runner_ref.add_event_listener(&input, "input", on_input)?;
+        runner_ref.add_event_listener(&input, "compositionstart", on_composition_start)?;
+        runner_ref.add_event_listener(&input, "compositionupdate", on_composition_update)?;
+        runner_ref.add_event_listener(&input, "compositionend", on_composition_end)?;
 
-    body.append_child(&input)?;
+        Ok(Self {
+            input,
+            prev_ime_output: Default::default(),
+        })
+    }
 
-    Ok(())
-}
+    pub fn move_to(
+        &self,
+        ime: Option<egui::output::IMEOutput>,
+        canvas: &web_sys::HtmlCanvasElement,
+    ) -> Result<(), JsValue> {
+        // don't move the text agent on mobile
```

Comment:
why?

Reply:
The `<input>` is moved to allow IME "popups" to follow the text cursor, but there's no need for that on mobile. Updated the comment